File: SegmentTree/SegmentTree.cpp
/*
 * @key word:线段树
 * @已测试:HDU - 1754
 * @Author: hgz 
 * @Date: 2017-04-24 19:09:25 
 * @Last Modified by:   hgz 
 * @Last Modified time: 2017-04-24 19:09:25 
 */
#include <cstdio>
#include <algorithm>

using namespace std;

const int MAXN = 200009;

struct line
{
    int l;
    int r;
    int cmax;
} Tree[MAXN * 4];

void pushUp(int t)
{
    int x = t << 1;
    Tree[t].cmax = max(Tree[x + 1].cmax, Tree[x + 2].cmax);
}

void buildTree(int l, int r, int t)
{
    Tree[t].l = l;
    Tree[t].r = r;
    Tree[t].cmax = 0;
    if (l == r)
    {
        scanf("%d", &Tree[t].cmax);
        return;
    }
    int x = (l + r) >> 1;
    buildTree(l, x, t << 1 | 1);
    buildTree(x + 1, r, (t << 1) + 2);
    pushUp(t);
}

void updateTree(int v, int cn, int t)
{
    if (Tree[t].l == v && Tree[t].r == v)
    {
        Tree[t].cmax = cn;
        return;
    }
    int m = (Tree[t].l + Tree[t].r) >> 1;
    if (v <= m)
    {
        updateTree(v, cn, t << 1 | 1);
    }
    else
    {
        updateTree(v, cn, (t << 1) + 2);
    }
    pushUp(t);
}

int queryTree(int l, int r, int t)
{
    if (Tree[t].l == l && Tree[t].r == r)
    {
        return Tree[t].cmax;
    }
    int m = (Tree[t].l + Tree[t].r) >> 1;
    if (l > m)
    {
        return queryTree(l, r, (t << 1) + 2);
    }
    else if (r <= m)
    {
        return queryTree(l, r, t << 1 | 1);
    }
    else
    {
        int ret1 = queryTree(l, m, t << 1 | 1);
        int ret2 = queryTree(m + 1, r, (t << 1) + 2);
        return max(ret1, ret2);
    }
}

int main()
{
    int n, m;
    while (scanf("%d%d", &n, &m) != EOF)
    {
        buildTree(1, n, 0);
        while (m--)
        {
            char op[2];
            int a, b;
            scanf("%s%d%d", op, &a, &b);
            if (op[0] == 'Q')
            {
                printf("%d\n", queryTree(a, b, 0));
            }
            else
            {
                updateTree(a, b, 0);
            }
        }
    }
    return 0;
}

File: SegmentTree/SegmentTree.pas
{
 * @key word:线段树
 * @已测试:HDU - 1166
 * @Author: hgz 
 * @Date: 2017-04-24 19:12:50 
 * @Last Modified by:   hgz 
 * @Last Modified time: 2017-04-24 19:12:50 
}
const maxn=50000;
var
tree:array[1..maxn*6]of longint;
a:array[1..maxn]of longint;
n,wz,i,r1,r2,sl,j:longint;
ch:char;
st:string;

procedure build(v,l,r:longint);
var
mid:longint;
begin
	if l=r then begin
		tree[v]:=a[l];
		exit;
		end;
	mid:=(l+r)div 2;
	build(v+v,l,mid);
	build(v+v+1,mid+1,r);
	tree[v]:=tree[v+v]+tree[v+v+1];
end;

procedure modify(v,i,l,r,val:longint);
var
mid:longint;
begin
	if l=r then begin
		tree[v]:=tree[v]+val;
		exit;
		end;
	mid:=(l+r)div 2;
	if mid>=i then
		modify(v+v,i,l,mid,val)
		else
		modify(v+v+1,i,mid+1,r,val);
	tree[v]:=tree[v+v]+tree[v+v+1];
end;

function query(v,l,r,x,y:longint):longint;
var
mid:longint;
begin
	if (l=x)and(r=y) then 
		exit(tree[v]);
	mid:=(l+r)div 2;
	if (y<=mid) then 
		exit(query(v+v,l,mid,x,y))
		else
	if (x>mid) then 
		exit(query(v+v+1,mid+1,r,x,y));
	exit(query(v+v,l,mid,x,mid)+query(v+v+1,mid+1,r,mid+1,y));
end;

begin
readln(sl);
for j:=1 to sl do
begin
	fillchar(tree,sizeof(tree),0);
	readln(n);
	for i:=1 to n do 
		read(a[i]);
	readln;
	build(1,1,n);
	writeln('Case ',j,':');
	while true do 
	begin
		readln(st);
		if st[1]='E' then break;
		ch:=st[1];
		wz:=pos(' ',st);
		delete(st,1,wz);
		wz:=pos(' ',st);
		val(copy(st,1,wz-1),r1);
		delete(st,1,wz);
		val(st,r2);
		if ch='Q' then 
			writeln(query(1,1,n,r1,r2))
			else 
		if ch='S' then 
			modify(1,r1,1,n,-r2)
			else
		if ch='A' then 
			modify(1,r1,1,n,r2);
	end;
end;
end.